frontends/p4/typeChecking/typeUnification.cpp

```diff
@@ -450,9 +450,8 @@ bool TypeUnification::unify(const BinaryConstraint *constraint) {
         }
         return constraint->reportError(constraints->getCurrentSubstitution());
     } else if (dest->is<IR::Type_Declaration>() && src->is<IR::Type_Declaration>()) {
-        bool canUnify = typeid(dest) == typeid(src) && dest->to<IR::Type_Declaration>()->name ==
```

Comment:
This is a pre-existing p4c bug. `typeid(dest) == typeid(src)` is always true since both `dest` and `src` are of `IR::Type*` types and typeid of these expressions matches. Changing to `typeid(*dest) == typeid(*src)` would yield an error as it will consider generic substituted type and correspondent substitution to be separate types, when they are not. This change just removes always true condition. No test changes / failures.

Reply:
I definitely agree this was always-true conjunct. I am just wondering if this is a bug that we are not checking that the "kinds" of types match, but I would say it is probably safe to compare names only. It doubt there can be two entities with the same name and one declaring different kind of thing then the other (e.g. extern vs. parser), as you already have an example where the types does not match correctly, I agree with just dropping the conjunct.

Reply:
If we take the exiting testcase for this code, then checking the typeids yields to very confusing error message. As I noted – the ordinary type and the substituted generic type are different nodes, so the typeid check would fail. However, the diagnostics produces is confusing, IIRC – cannot unify "int" with "int" or something like this.